spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,133 @@
+  private static final String DELETE = ChangelogOperation.DELETE.name();
+  private static final String INSERT = ChangelogOperation.INSERT.name();
+
+  private final Iterator<Row> rowIterator;
+  private final int changeTypeIndex;
+  private final List<Integer> partitionIdx;
+  private final boolean markUpdatedRows;
+
+  private Row cachedRow = null;
+
+  public ChangelogIterator(
+      Iterator<Row> rowIterator,
+      int changeTypeIndex,
+      List<Integer> partitionIdx,
+      boolean markUpdatedRows) {
+    this.rowIterator = rowIterator;
+    this.changeTypeIndex = changeTypeIndex;
+    this.partitionIdx = partitionIdx;
+    this.markUpdatedRows = markUpdatedRows;
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (cachedRow != null) {
+      return true;
+    }
+    return rowIterator.hasNext();
+  }
+
+  @Override
+  public Row next() {
+    // if there is a processed cached row, return it directly
+    if (cachedRow != null
+        && !cachedRow.getString(changeTypeIndex).equals(DELETE)
+        && !cachedRow.getString(changeTypeIndex).equals(INSERT)) {
+      Row row = cachedRow;
+      cachedRow = null;
+      return row;
+    }
+
+    Row currentRow = currentRow();
+
+    if (rowIterator.hasNext()) {
+      GenericRowWithSchema nextRow = (GenericRowWithSchema) rowIterator.next();
+
+      if (withinPartition(currentRow, nextRow)
+          && currentRow.getString(changeTypeIndex).equals(DELETE)
+          && nextRow.getString(changeTypeIndex).equals(INSERT)) {
+
+        GenericInternalRow deletedRow =
+            new GenericInternalRow(((GenericRowWithSchema) currentRow).values());
+        GenericInternalRow insertedRow = new GenericInternalRow(nextRow.values());
+
+        // set the change_type to the same value
+        deletedRow.update(changeTypeIndex, "");
+        insertedRow.update(changeTypeIndex, "");
+
+        if (deletedRow.equals(insertedRow)) {
+          // remove two carry-over rows
+          currentRow = null;
+          this.cachedRow = null;
+        } else if (markUpdatedRows) {
+          // mark the updated rows
+          deletedRow.update(changeTypeIndex, ChangelogOperation.UPDATE_BEFORE.name());
+          currentRow = RowFactory.create(deletedRow.values());
+
+          insertedRow.update(changeTypeIndex, ChangelogOperation.UPDATE_AFTER.name());
+          this.cachedRow = RowFactory.create(insertedRow.values());
+        } else {
+          // recover the values of change type
+          deletedRow.update(changeTypeIndex, DELETE);
+          insertedRow.update(changeTypeIndex, INSERT);
+          this.cachedRow = nextRow;
+        }
+
+      } else {
+        this.cachedRow = nextRow;
+      }
+    }
+
+    return currentRow;
+  }
+
+  private Row currentRow() {
+    if (cachedRow != null) {
+      Row row = cachedRow;
+      cachedRow = null;
+      return row;
+    } else {
+      return rowIterator.next();
+    }
+  }
+
+  private boolean withinPartition(Row currentRow, Row nextRow) {
```

Comment:
isn't this "sameRow" basically?  Since we are basically checking are these two rows are and pre and post update correct?

Reply:
Yes. The partition columns here are identifier columns. This iterator will only be used in CDC procedure, so it is safe to assume that.

Reply:
These two rows can be a update row or a carry-over row.

Reply:
Are we ok with the name 'partition'?  Wont it be overloaded in Iceberg?

Reply:
It is overloaded. Here the `partition` refer to the one generated by `DataFrame::repartition()`. I'm not sure what would be a better name. Open for suggestions.

Reply:
Yea tbh it threw me off a bit initially.  How about just 'equivalent'?  And partitionFieldIdx => identifierFieldIdx?  Not sure what you guys think.

Reply:
I think something like "sameLogicalRow(Row currentRow, Row nextRow)"